students/john_rogers/lesson04/mailroom_pt2.py

```diff
@@ -0,0 +1,183 @@
+                   },
+          'santa': {'d1': 45.43,
+                    'd2': 98345.32,
+                    'd3': 54.34,
+                    'd4': 456.23,
+                    },
+         }
+
+    # create the main user prompt
+    main_prompt = (
+        "\nWelcome to the main menu!\n"
+        "Please pick a number from the following:\n"
+        "1: exit the program\n"
+        "2: check donor list and become a donor\n"
+        "3: create an interactive report of current donors\n"
+        "4: write a thank you note for each donor and write to file\n"
+        ">>>\n"
+    )
+
+    # create the main menu using dict switch
+    main_dispatch = {
+        '1': exit_menu,
+        '2': thank_you,
+        '3': create_report,
+        '4': write_report,
+    }
+
+    # calling menu with no variables at this point since they are global
+    menu()
+
+
+def menu():
+    """
+    Get user input
+    :return: call the appropriate menu item
+    """
+    while True:
+        response = input(main_prompt)
+        if response == '1':
+            exit_menu()
+        else:
+            main_dispatch[response](db)
+
+
+def write_report():
+    pass
+
+
+def exit_menu():
+    """
+    exit the program
+    :return: SystemExit
+    """
+    print('Exiting program -')
+    raise SystemExit
+
+
+def thank_you(names):
+    """
+    send a thank you, check the donor list or add donation
+    :return: None
+    """
+
+    while True:
+        print('Enter q to exit to main menu.')
+        cmd = input("Enter 'list' to see a current list of donors or "
+                    "a new name to become a new donor: ")
+        cmd = cmd.lower()
+
+        # TODO: v-next decompose this into smaller, more discrete functions
+        # TODO: validate donations are digits instead of strings
+        if cmd.isdigit():
+            break
+        elif cmd == 'q':
+            break
+        elif cmd == 'list':
+            print()
+            for i in names.keys():
+                print(i.capitalize())
+        elif cmd in names.keys():
+            # name already in list, solicit new donation and add to dict
+            donation = input('Please enter an amount to donate: ')
+            donation = float(donation)
+
+            d_num = len(names[cmd].keys()) + 1
+            d_num = str(d_num)
+            new_key = 'd' + d_num
+            names[cmd][new_key] = donation
+
+            print(f'Thank you, {cmd.capitalize()}, for your kind'
+                  f' donation of ${donation}.')
+        else:
+            # add the new name and prompt for donation
+            print(f'Welcome aboard, {cmd.capitalize()}, how much would '
+                  f'you like to donate?')
+            new_donation = input('Please an amount to donate: ')
+            new_donation = float(new_donation)
+            names[cmd] = {'d1': new_donation}
+
+            print(f'Thank you, {cmd.capitalize()}, for your kind'
+                  f' donation of ${new_donation}.')
```

Comment:
there is quite a bit of duplicate code in the lines 128 to 149 - I would recommend refactoring that.